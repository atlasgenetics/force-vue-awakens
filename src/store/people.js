const initialState = () => ({
  page: 1,
  next: null,
  previous: null,
  results: []
})

// Export the module so it can be included in the main store.
export default {
  // This module is namespaced so that it is a little more self-contained:
  // https://vuex.vuejs.org/guide/modules.html#namespacing
  namespaced: true,
  // https://vuex.vuejs.org/guide/state.html
  state: initialState,
  // https://vuex.vuejs.org/guide/mutations.html
  mutations: {
    // Save results and next/previous links to the store
    results: (state, response) => {
      state.results = response.results
      state.next = response.next
      state.previous = response.previous
    },
    // Update the current page number
    page: (state, page) => (state.page = page)
  },
  // https://vuex.vuejs.org/guide/actions.html
  actions: {
    // This action takes state to use the current page as a default value
    async list ({ commit, state }, page = state.page) {
      // Fetch People from SWAPI and await the response.
      const response = await fetch('https://swapi.co/api/people/?page=' + page)
      if (response.ok) {
        // If the response is valid, parse the JSON string and commit the result
        commit('results', await response.json())
        // Commit the new page number
        commit('page', page)
      } else {
        // If the response is not OK, log the response to the console.
        console.error(response)
      }
    }
  }
}
