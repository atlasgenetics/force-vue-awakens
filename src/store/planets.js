const initialState = () => ({
  results: []
})

// Export the module so it can be included in the main store.
export default {
  // This module is namespaced so that it is a little more self-contained:
  // https://vuex.vuejs.org/guide/modules.html#namespacing
  namespaced: true,
  // https://vuex.vuejs.org/guide/state.html
  state: initialState,
  // https://vuex.vuejs.org/guide/mutations.html
  mutations: {
    results: (state, results = []) => (state.results = results)
  },
  // https://vuex.vuejs.org/guide/actions.html
  actions: {
    async listPlanets ({ commit }, page = 1) {
      const response = await fetch('https://swapi.co/api/planets')
      const { results } = await response.json()
      if (response.ok) {
        // If the response is OK commit the data to the store.
        commit('results', results)
      } else {
        // If the response is not OK, log the response to the console.
        console.error(response)
      }
    }
  },
  getters: {
    planets: state => state.results
  }
}
