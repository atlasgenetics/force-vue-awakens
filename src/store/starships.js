export default{

  namespaced: true,

  state: {
    starships: [],
    shipInfo: [],
    shipPilots: [],
    shipFilms: []
  },
  mutations: {
    listStarships(state, data){
      state.starships = data.map((el) => {
        //get starship id from url
        let id = el.url.match(/([0-9])+/g)
        //add id to starship object
        el.id = id[0];
        return el
      })
    },
    setShipInfo(state, data){
      //set ship data
      state.shipInfo = data
      //clear previous pilots and films
      state.shipPilots = []
      state.shipFilms = []
    },
    addShipFilm(state, data){
      state.shipFilms.push({id: data.id, title: data.title, release_date: data.release_date})
    },
    addShipPilot(state, data){
      state.shipPilots.push({id: data.id, name: data.name})
    }
  },
  actions: {
    async list({commit}){
       //get starships from SWAPI API
       const response  = await fetch('https://swapi.co/api/starships/')
       //Parse JSON response
       const { results } = await response.json()
       if(response.ok){
         commit('listStarships', results)
       } else {
         //if response is not 'OK' log it to the console
         console.log(response)
       }
    },
    async get({commit, dispatch}, id){
      //get ship info from SWAPI API
      const response = await fetch('https://swapi.co/api/starships/'+ id + '/')
      //Parse JSON response
      const result = await response.json()
      if(response.ok){
        commit('setShipInfo', result)
        //set film values if applicable
        result.films.forEach((val, key) => {
            dispatch('getResource',{url: val, typ: 'films'})
        })
        //set pilot values if applicable
        result.pilots.forEach((val, key) => {
          dispatch('getResource',{url: val, typ: 'people'})
        })
      } else{
        //if response is not 'OK' log it to the console
        console.log(response)
      }
    },
    async getResource({commit}, data){
      //get resource id from url
      const id = data.url.match(/([0-9])+/g)
      //get resource info from SWAPI API
      const response = await fetch('https://swapi.co/api/'+ data.typ +'/'+ id[0] + '/')
      //Parse JSON response
      const result = await response.json()
      if(response.ok){
        //run typ specific mutation
        switch(data.typ){
          case 'films':
            commit('addShipFilm', {id: id[0], title: result.title, release_date: result.release_date})
          break;
          case 'people':
            commit('addShipPilot', {id: id[0], name: result.name})
          break;
        }
      } else{
        //if response is not 'OK' log it to the console
        console.log(response)
      }
    }
  }
}
