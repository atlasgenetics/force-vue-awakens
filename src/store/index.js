import Vue from 'vue'
import Vuex from 'vuex'

import people from '@/store/people'
import starships from '@/store/starships'

// Instruct Vue to use the Vuex plugin.
Vue.use(Vuex)

// Export a new Vuex Store instance to be used in the application.
export default new Vuex.Store({
  // Modules are mini-state stores that can be combined into the main state
  // tree: https://vuex.vuejs.org/guide/modules.html
  modules: {
    people,
    starships
  }
})
