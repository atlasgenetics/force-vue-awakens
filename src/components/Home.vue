<template>
  <div class="container pt-4 pb-4">

    <section>

      <h2>People</h2>

      <div class="d-flex flex-wrap">
        <router-link
          v-for="(person, id) in results"
          :to="`/people/${id}`"
          :key="id"
          class="card m-2"
          style="width: 12rem;">

          <img
            :src="`https://starwars-visualguide.com/assets/img/characters/${id + 1}.jpg`"
            class="card-img-top"
            alt="Card image cap">

          <div class="card-body">

            <h5 class="card-title m-0">
              {{ person.name }}
            </h5>

          </div>

        </router-link>
      </div>

    </section>
    <section>

      <h2>Starships</h2>

      <div class="d-flex flex-wrap">
        <router-link
          v-for="starship in starships"
          :to="`/starship/${starship.id}`"
          :key="starship.id"
          class="card m-2"
          style="width: 12rem;">

          <img
            :src="`https://starwars-visualguide.com/assets/img/starships/${starship.id}.jpg`"
            class="card-img-top"
            alt="Card image cap">

          <div class="card-body">

            <h5 class="card-title m-0">
              {{ starship.name }}
            </h5>

          </div>

        </router-link>
      </div>

    </section>

  </div>
</template>

<script>
import { mapState } from 'vuex'

import store from '@/store'

export default {
  async beforeRouteEnter(to, from, next) {
    await store.dispatch('people/list')
    await store.dispatch('starships/list')
    next()
  },
  computed: {
    ...mapState('people', ['results']),
    ...mapState('starships', ['starships'])
  }
}
</script>
