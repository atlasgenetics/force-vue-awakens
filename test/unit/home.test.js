import { mount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'

import store from '@/store'
import Home from '@/components/Home'

const localVue = createLocalVue()

localVue.use(Vuex)

test('Home renders correctly', () => {
  const wrapper = mount(Home, {
    store,
    localVue,
    methods: {
      errorTest: () => {}
    }
  })
  expect(wrapper.html()).toMatchSnapshot()
})
