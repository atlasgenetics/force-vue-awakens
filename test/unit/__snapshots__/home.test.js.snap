// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Home renders correctly 1`] = `
"<div class=\\"container pt-4 pb-4\\">
  <section>
    <h2>People</h2>
    <div class=\\"d-flex flex-wrap\\"></div>
  </section>
</div>"
`;
